### ISSUE TITLE
Windows 10 does not notify of disconnection using USB serial devices

### ISSUE THREAD

**Issue by stephanemagnenat** [aseba-community/dashel#22]
On Windows versions before 10, using Thymio provided USB serial (rfcomm) driver, the disconnection of the physical device leads to a write failure, leading Dashel to be aware of the disconnection.

On Windows 10, which provides its own `rfcomm` driver, *the physical disconnection of the device does not lead to a write failure*, so Dashel cannot realise that the disconnection happened.

This is not a bug in Dashel, but in the Windows 10 driver. There might be a work-around by creating a hidden windows that listens to the system event for the USB device itself, but it is super-complicated.

**Comment by ypiguet-epfl** [comment]
I've tried to find a work-around. Please have a look at this [c++ file](https://github.com/aseba-community/dashel/files/1000318/portlist.txt), a modified version of examples/portmap.cpp I've compiled in Visual Studio Community 2017. It contains suggestions for Dashel.

**Comment by stephanemagnenat** [comment]
This is very interesting, I did not notice the timeout settings in the serial port. So in your approach, if the serial port timeouts, either the device has been disconnected and we through a Dashel disconnected exception, or it is still connected, and we try again until the next timeout, right?

About changing `SerialPortEnumerator::getPorts()`, I am not sure that it is the best approach, as the serial port passed in the target might be set by the user might be a name, a port, or a device (see http://aseba-community.github.io/dashel/). Rather, I would get the correct `DevInst` for the corresponding serial port in the constructor of `SerialStream`, and store it in that class.

**Comment by ypiguet-epfl** [comment]
(i) Yes. I don't know why such large timeout values were selected (100 s in cto in dashel-win32.cpp); maybe they weren't needed when there was a simpler way to be notified about disconnections. I haven't installed the environment required to compile Aseba Studio, so my tests have been limited.
(ii) Yes, whatever makes sense and allows a clean implementation.

**Comment by stephanemagnenat** [comment]
> (i) Yes. I don't know why such large timeout values were selected (100 s in cto in dashel-win32.cpp); maybe they weren't needed when there was a simpler way to be notified about disconnections.

At the time of writing Dashel, physical disconnections lead to read/write errors. But it was 10 years ago.

**Comment by mbonani** [comment]
After discussing with Jonathan (Windows specialist), I think the solution propose by Yves has a good feasibility. @ypiguet-epfl could you implement it? Can I help in installing a Windows built or for testing?

**Comment by ypiguet-epfl** [comment]
Ok. Thanks for your help offer! I'll try on my own unless I have a problem.

**Comment by stephanemagnenat** [comment]
@ypiguet-epfl note that there is detailed instruction on how to setup the environment here: https://github.com/aseba-community/aseba/blob/master/compile.Windows.md

**Comment by ypiguet-epfl** [comment]
I'm struggling with the first step, reducing the timeout and know there is a problem. I've tried to reduce the read timeout in `COMMTIMEOUTS` (dashel-win32.cpp 876:878), but there must be something I don't understand in the overlapped I/O in `FileStream::read`.
But anyway, Aseba Studio must already detect a timeout, since when the connection is broken the left part of the window (list of variables etc.) is hidden. Why not assuming a disconnection, closing the `SerialStream` and displaying the "Connection closed, I'm trying to reconnect" dialog? That would solve the problem...

**Comment by stephanemagnenat** [comment]
> Why not assuming a disconnection, closing the SerialStream and displaying the "Connection closed, I'm trying to reconnect" dialog? That would solve the problem...

That's not possible because it is not the same protocol layer. Aseba Studio detects the disconnection per node, as this can happen in a wireless network. This is different than a Dashel disconnection.

**Comment by ypiguet-epfl** [comment]
What's possible then? Killing the unresponsive Dashel stream from Studio? or the whole Dashel hub? A new intermediate layer? Redesigning Dashel and/or Studio and/or the Thymio firmware?

**Comment by stephanemagnenat** [comment]
I think that the main question is where the thread using the Dashel Hub is blocked upon disconnection. My expectation is that it is in `Hub::step()`. So this is the first thing to validate.

**If it is the case**, no need to change the architecture. I think the solution can be based on one of your early experiment in which have a small program that enumerates the serial ports every one second. Here is my proposal:

1. Having a thread _serialcheck_ that keeps track of which physical USB devices are being used for USB serial connections, and every second poll if these devices are still present.
2. When a USB serial connection is created, take note of which physical device it is attached to and inform _serialcheck_.
3. When a USB serial connection is closed, inform _serialcheck_.
4. If _serialcheck_ detects that one of the tracked device has been removed, inform the Hub thread through a shared Event. This mechanism is already used for `stop()` through the event `hTerminate`. Use this event inside the core call to `WaitForMultipleObjects` at line 1408.
5. One we get out of `WaitForMultipleObjects`, check whether we exited because of the _serialcheck_ event. If so, iterate over all serial connections to see which one must be closed. Close it normally.

**If it is not the case**, we have to understand better exactly what happens.
